In [1]:
# importing pandas as pd
import pandas as pd

In [2]:
# reading ExcelFile.
xls = pd.ExcelFile('Data.xlsx')

# preview of sheet names.
xls.sheet_names

['RH K-4 MOY Data',
 'MOY Export',
 'Syracuse',
 'Howard',
 'Swarthmore',
 'Spelman',
 'Williams',
 'Brooklyn',
 'Tuskegee',
 'Northwestern',
 'Georgetown',
 'Wisconsin']

In [3]:
df = pd.concat(pd.read_excel(xls, sheet_name=xls.sheet_names[2:]), ignore_index=True)
df.shape
df.head()

Student_ Number   Grade_Level  Home_room  BOY Indep Level  \
0       236563003.0           0.0   Syracuse              AA   
1       238194344.0           0.0   Syracuse              AA   
2       239980246.0           0.0   Syracuse              AA   
3       240845834.0           0.0   Syracuse              AA   
4       239980196.0           0.0   Syracuse              AA   

    Grade Level Equiv  Inst Level  Grade Level Equiv.1 MOY Indep Level  \
0               -0.25         NaN                                   AA   
1               -0.25         NaN                                   AA   
2               -0.25         NaN                                   AA   
3               -0.25         NaN                                   AA   
4               -0.25         NaN                                   AA   

   Grade Level Equiv.2  Inst Level.1  Grade Level Equiv.3   Accuracy (%)  \
0                -0.25             A                  0.0           88.0   
1                -0.25             A                  0.0           97.0   
2                -0.25             A                  0.0           33.0   
3                                 AA                -0.25            NaN   
4                                 AA                -0.25            NaN   

    Fluency (0-3)   Comp Within (0-3)   Comp Beyond (0-3)  Comp About  \
0             1.0                 3.0                 2.0         NaN   
1             2.0                 2.0                 1.0         NaN   
2             2.0                 1.0                 2.0         NaN   
3             3.0                 1.0                 NaN         NaN   
4             2.0                 3.0                 1.0         NaN   

    Total Comp  
0          5.0  
1          3.0  
2          3.0  
3          1.0  
4          4.0

In [4]:
# previewing df
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 274 entries, 0 to 273
Data columns (total 17 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   Student_ Number       274 non-null    float64
 1   Grade_Level           274 non-null    float64
 2   Home_room             274 non-null    object 
 3   BOY Indep Level       273 non-null    object 
 4    Grade Level Equiv    259 non-null    float64
 5    Inst Level           187 non-null    object 
 6    Grade Level Equiv.1  274 non-null    object 
 7   MOY Indep Level       271 non-null    object 
 8    Grade Level Equiv.2  274 non-null    object 
 9    Inst Level.1         264 non-null    object 
 10   Grade Level Equiv.3  254 non-null    object 
 11   Accuracy (%)         218 non-null    float64
 12   Fluency (0-3)        232 non-null    float64
 13   Comp Within (0-3)    251 non-null    float64
 14   Comp Beyond (0-3)    249 non-null    float64
 15   Comp About           1

A couple of noticings:

* The column names are inconsistently formatted and need to be cleaned.
* Some numerical columns need to be casted as such.
* Since the Non-null values are not consistent, further investigation is needed.

In [5]:
# removing spaces and special characters from column names 
df.columns = df.columns.str.strip()
df.columns = df.columns.str.replace('_',' ')

df.insert(1,'Student Number', df['Student  Number'])
df.insert(2,'Homeroom', df['Home room'])

# columns to drop
drop_cols = ['Home room','Student  Number']
# drop columns
df.drop(drop_cols, axis=1, inplace=True)

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 274 entries, 0 to 273
Data columns (total 17 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Student Number       274 non-null    float64
 1   Homeroom             274 non-null    object 
 2   Grade Level          274 non-null    float64
 3   BOY Indep Level      273 non-null    object 
 4   Grade Level Equiv    259 non-null    float64
 5   Inst Level           187 non-null    object 
 6   Grade Level Equiv.1  274 non-null    object 
 7   MOY Indep Level      271 non-null    object 
 8   Grade Level Equiv.2  274 non-null    object 
 9   Inst Level.1         264 non-null    object 
 10  Grade Level Equiv.3  254 non-null    object 
 11  Accuracy (%)         218 non-null    float64
 12  Fluency (0-3)        232 non-null    float64
 13  Comp Within (0-3)    251 non-null    float64
 14  Comp Beyond (0-3)    249 non-null    float64
 15  Comp About           132 non-null    obj

In [7]:
# changing data to numerical type
df[['Grade Level Equiv.1','Grade Level Equiv.2','Grade Level Equiv.3']] = df[['Grade Level Equiv.1','Grade Level Equiv.2','Grade Level Equiv.3']].apply(pd.to_numeric, errors='coerce')

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 274 entries, 0 to 273
Data columns (total 17 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Student Number       274 non-null    float64
 1   Homeroom             274 non-null    object 
 2   Grade Level          274 non-null    float64
 3   BOY Indep Level      273 non-null    object 
 4   Grade Level Equiv    259 non-null    float64
 5   Inst Level           187 non-null    object 
 6   Grade Level Equiv.1  173 non-null    float64
 7   MOY Indep Level      271 non-null    object 
 8   Grade Level Equiv.2  234 non-null    float64
 9   Inst Level.1         264 non-null    object 
 10  Grade Level Equiv.3  248 non-null    float64
 11  Accuracy (%)         218 non-null    float64
 12  Fluency (0-3)        232 non-null    float64
 13  Comp Within (0-3)    251 non-null    float64
 14  Comp Beyond (0-3)    249 non-null    float64
 15  Comp About           132 non-null    obj

In [9]:
# get summary of missing values
df.isna().sum()

Student Number           0
Homeroom                 0
Grade Level              0
BOY Indep Level          1
Grade Level Equiv       15
Inst Level              87
Grade Level Equiv.1    101
MOY Indep Level          3
Grade Level Equiv.2     40
Inst Level.1            10
Grade Level Equiv.3     26
Accuracy (%)            56
Fluency (0-3)           42
Comp Within (0-3)       23
Comp Beyond (0-3)       25
Comp About             142
Total Comp               0
dtype: int64

For BOY and MOY missing data, I will use the key provided with this analysis.

In [10]:
# importing BOY info
sheet_url = 'https://docs.google.com/spreadsheets/d/1ctQdLgknVDQ5BPCS-O6UihzK5kPbbVwI-svT3cWKWfw/edit#gid=1026408630'
url1 = sheet_url.replace('/edit#gid=', '/export?format=csv&gid=')        
df_levels = pd.read_csv(url1, header=1)
df_levels.head(27)

Independent Reading Level  Grade Equivalent                  K  \
0                         AA             -0.25     On Grade Level   
1                          A              0.00  Above Grade Level   
2                          B              0.33  Above Grade Level   
3                          C              0.67  Above Grade Level   
4                          D              1.00  Above Grade Level   
5                          E              1.16  Above Grade Level   
6                          F              1.33  Above Grade Level   
7                          G              1.50  Above Grade Level   
8                          H              1.67  Above Grade Level   
9                          I              1.83  Above Grade Level   
10                         J              2.00  Above Grade Level   
11                         K              2.33  Above Grade Level   
12                         L              2.67  Above Grade Level   
13                         M              3.00  Above Grade Level   
14                         N              3.33  Above Grade Level   
15                         O              3.67  Above Grade Level   
16                         P              4.00  Above Grade Level   
17                         Q              4.33  Above Grade Level   
18                         R              4.67  Above Grade Level   
19                         S              5.00  Above Grade Level   
20                         T              5.33  Above Grade Level   
21                         U              5.67  Above Grade Level   
22                         V              6.00  Above Grade Level   
23                         W              6.33  Above Grade Level   
24                         X              6.67  Above Grade Level   
25                         Y              7.00  Above Grade Level   
26                         Z              8.00  Above Grade Level   

                          1                        2                        3  \
0         Below Grade Level        Below Grade Level        Below Grade Level   
1         Below Grade Level        Below Grade Level        Below Grade Level   
2         Below Grade Level        Below Grade Level        Below Grade Level   
3   Approaching Grade Level        Below Grade Level        Below Grade Level   
4            On Grade Level        Below Grade Level        Below Grade Level   
5         Above Grade Level        Below Grade Level        Below Grade Level   
6         Above Grade Level        Below Grade Level        Below Grade Level   
7         Above Grade Level        Below Grade Level        Below Grade Level   
8         Above Grade Level        Below Grade Level        Below Grade Level   
9         Above Grade Level  Approaching Grade Level        Below Grade Level   
10        Above Grade Level           On Grade Level        Below Grade Level   
11        Above Grade Level        Above Grade Level        Below Grade Level   
12        Above Grade Level        Above Grade Level  Approaching Grade Level   
13        Above Grade Level        Above Grade Level           On Grade Level   
14        Above Grade Level        Above Grade Level        Above Grade Level   
15        Above Grade Level        Above Grade Level        Above Grade Level   
16        Above Grade Level        Above Grade Level        Above Grade Level   
17        Above Grade Level        Above Grade Level        Above Grade Level   
18        Above Grade Level        Above Grade Level        Above Grade Level   
19        Above Grade Level        Above Grade Level        Above Grade Level   
20        Above Grade Level        Above Grade Level        Above Grade Level   
21        Above Grade Level        Above Grade Level        Above Grade Level   
22        Above Grade Level        Above Grade Level        Above Grade Level   
23        Above Grade Level        Above Grade Level        Above Grade Level   
24        Above Grade Level        Above Grad

In [11]:
# Importing MOY info
sheet_url = 'https://docs.google.com/spreadsheets/d/1ctQdLgknVDQ5BPCS-O6UihzK5kPbbVwI-svT3cWKWfw/edit#gid=1314641425'
url2 = sheet_url.replace('/edit#gid=', '/export?format=csv&gid=')        
df_MOY = pd.read_csv(url2, header=1)
df_MOY.head(27)

Independent Reading Level  Grade Equivalent                        K  \
0                         AA             -0.25        Below Grade Level   
1                          A              0.00  Approaching Grade Level   
2                          B              0.33           On Grade Level   
3                          C              0.67        Above Grade Level   
4                          D              1.00        Above Grade Level   
5                          E              1.16        Above Grade Level   
6                          F              1.33        Above Grade Level   
7                          G              1.50        Above Grade Level   
8                          H              1.67        Above Grade Level   
9                          I              1.83        Above Grade Level   
10                         J              2.00        Above Grade Level   
11                         K              2.33        Above Grade Level   
12                         L              2.67        Above Grade Level   
13                         M              3.00        Above Grade Level   
14                         N              3.33        Above Grade Level   
15                         O              3.67        Above Grade Level   
16                         P              4.00        Above Grade Level   
17                         Q              4.33        Above Grade Level   
18                         R              4.67        Above Grade Level   
19                         S              5.00        Above Grade Level   
20                         T              5.33        Above Grade Level   
21                         U              5.67        Above Grade Level   
22                         V              6.00        Above Grade Level   
23                         W              6.33        Above Grade Level   
24                         X              6.67        Above Grade Level   
25                         Y              7.00        Above Grade Level   
26                         Z              8.00        Above Grade Level   

                          1                        2                        3  \
0         Below Grade Level        Below Grade Level        Below Grade Level   
1         Below Grade Level        Below Grade Level        Below Grade Level   
2         Below Grade Level        Below Grade Level        Below Grade Level   
3         Below Grade Level        Below Grade Level        Below Grade Level   
4         Below Grade Level        Below Grade Level        Below Grade Level   
5         Below Grade Level        Below Grade Level        Below Grade Level   
6   Approaching Grade Level        Below Grade Level        Below Grade Level   
7            On Grade Level        Below Grade Level        Below Grade Level   
8         Above Grade Level        Below Grade Level        Below Grade Level   
9         Above Grade Level        Below Grade Level        Below Grade Level   
10        Above Grade Level  Approaching Grade Level        Below Grade Level   
11        Above Grade Level           On Grade Level        Below Grade Level   
12        Above Grade Level        Above Grade Level        Below Grade Level   
13        Above Grade Level        Above Grade Level  Approaching Grade Level   
14        Above Grade Level        Above Grade Level           On Grade Level   
15        Above Grade Level        Above Grade Level        Above Grade Level   
16        Above Grade Level        Above Grade Level        Above Grade Level   
17        Above Grade Level        Above Grade Level        Above Grade Level   
18        Above Grade Level        Above Grade Level        Above Grade Level   
19        Above Grade Level        Above Grade Level        Above Grade Level   
20        Above Grade Level        Above Grade Level        Above Grade Level   
21        Above Grade Level        Above Grade Level        Above Grade Level   
22        Above Grade Level        Abov

BOY and MOY Independent Reading Levels & Grade Equivalents are identical.  I will create a dictionary with these level to assist with missing values.

In [13]:
# Dictionary 
alpha_keys = list(df_levels['Independent Reading Level'].unique())
alpha_values = list(df_levels['Grade Equivalent'].unique())

levels = dict(zip(alpha_keys, alpha_values))
levels

{'AA': -0.25,
 'A': 0.0,
 'B': 0.33,
 'C': 0.67,
 'D': 1.0,
 'E': 1.16,
 'F': 1.33,
 'G': 1.5,
 'H': 1.67,
 'I': 1.83,
 'J': 2.0,
 'K': 2.33,
 'L': 2.67,
 'M': 3.0,
 'N': 3.33,
 'O': 3.67,
 'P': 4.0,
 'Q': 4.33,
 'R': 4.67,
 'S': 5.0,
 'T': 5.33,
 'U': 5.67,
 'V': 6.0,
 'W': 6.33,
 'X': 6.67,
 'Y': 7.0,
 'Z': 8.0}

In [15]:
# locating records with missing BOY Grade Equivalent data
print(df.loc[df['Grade Level Equiv'].isna()].shape)
df.loc[df['Grade Level Equiv'].isna()]

(15, 17)


Student Number      Homeroom  Grade Level BOY Indep Level  \
24      240415372.0      Syracuse          0.0                   
26      239668965.0      Syracuse          0.0                   
53      240905430.0        Howard          0.0                   
60      205888340.0    Swarthmore          1.0                   
67      238231914.0    Swarthmore          1.0             NaN   
73      237499462.0    Swarthmore          1.0                   
82      234292928.0    Swarthmore          1.0                   
83      218702165.0    Swarthmore          1.0                   
138     234743243.0      Williams          2.0                   
139     234411106.0      Williams          2.0                   
175     226787604.0      Tuskegee          3.0                   
189     233199898.0      Tuskegee          3.0                   
190     231197146.0      Tuskegee          3.0                   
195     203306691.0  Northwestern          3.0               H   
230     232795195.0    Georgetown          4.0                   

     Grade Level Equiv Inst Level  Grade Level Equiv.1 MOY Indep Level  \
24                 NaN                             NaN              AA   
26                 NaN                             NaN              AA   
53                 NaN                             NaN                   
60                 NaN                             NaN               C   
67                 NaN        NaN                  NaN               D   
73                 NaN                             NaN               F   
82                 NaN                             NaN                   
83                 NaN                             NaN                   
138                NaN                             NaN                   
139                NaN                             NaN                   
175                NaN                             NaN               M   
189                NaN                             NaN                   
190                NaN                             NaN                   
195                NaN                             NaN               I   
230                NaN                             NaN               Q   

     Grade Level Equiv.2 Inst Level.1  Grade Level Equiv.3  Accuracy (%)  \
24                   NaN            A                  NaN           NaN   
26                   NaN            A                  NaN           NaN   
53                   NaN                               NaN           NaN   
60                  0.67            D                 1.00          99.0   
67                  1.00            E                 1.16           NaN   
73                  1.33            G                 1.50          94.0   
82                   NaN                               NaN           NaN   
83                   NaN                               NaN           NaN   
138                  NaN                               NaN           NaN   
139                  NaN                               NaN           NaN   
175                 3.00            N                  NaN          99.0   
189                  NaN                               NaN           NaN   
190                  NaN                               NaN           NaN   
195                  NaN            J                  NaN          98.0   
230                 4.33            R                 4.67          99.0   

     Fluency (0-3)  Comp Within (0-3)  Comp Beyond (0-3) Comp About  \
24             NaN                NaN                NaN              
26             NaN                NaN                NaN              
53             NaN                NaN                NaN              
60             2.0                2.0                2.0        NaN   
67             1.0                1.0                2.0        NaN   
73             1.0                NaN                1.0        NaN   
82             NaN                NaN            

Upon further investigation, only one BOY Independent Level was included (level H). My assumption is that these students do not have beginning year data because they enrolled in school after school had started. I will address the missing equivalency for the one student with data next. Afterwards, I will fill those values with "Not Enrolled." 

In [ ]:
# replacing grade equivalence for level H
df.loc[(df['BOY Indep Level']=='H'),'Grade Level Equiv']=1.67

# checking to see if change was made.
df.loc[df['Grade Level Equiv'].isna()].shape

In [ ]:
# Locating records with missing MOY Grade Equivalent data
print(df.loc[df['Grade Level Equiv.2'].isna()].shape)
df.loc[df['Grade Level Equiv.2'].isna()]

In [ ]:
# Locating records with missing BOY Inst Level data
df.loc[df['Inst Level'].isna()]

In [ ]:
df['Grade Level Equiv'].max()

The data shows that the instructional level is typically one level above the BOY Indep Level. 5.67 also appears to be highest level, which is equivalent to level U.

In [ ]:
# Filling Instructional Level given BOY Independent Levels
i = 0
for i in range(len(alpha_keys)-1):
    for level in alpha_keys:
        df.loc[(df['BOY Indep Level']==alpha_keys[i]), 'Inst Level']=alpha_keys[i+1]
        i=+1

In [ ]:
# Filling null values:
df['Grade Level Equiv'].fillna('Not Enrolled', inplace=True)

# Verifying the filling functioned correctly.
df.loc[df['Grade Level Equiv'].isna()]

df.loc[(df['Grade Level Equiv']=='Not Enrolled'), 'BOY Indep Level']='Not Enrolled'
df.loc[(df['Grade Level Equiv']=='Not Enrolled'), 'Inst Level']='Not Enrolled'
df.loc[(df['Grade Level Equiv']=='Not Enrolled'), 'Grade Level Equiv']='Not Enrolled'
df.loc[(df['Grade Level Equiv']=='Not Enrolled'), 'Grade Level Equiv.1']='Not Enrolled'

df.loc[df['Grade Level Equiv']=='Not Enrolled']

In [ ]:
df.loc[df['Inst Level'].notnull()]

In [ ]:
# Addressing Grade Level Equiv.1 for Instructional Level
for level, equiv in levels.items():
    df.loc[(df['Inst Level']==level), 'Grade Level Equiv.1'] = equiv
    
# Checking for accuracy    
df.loc[df['Inst Level'].notnull()]

In [ ]:
# Locating records with missing MOY Indepent Level data
df.loc[df['MOY Indep Level'].isna()]

The Mid-Year Instructional Level is included for students with missing MOY Independent Levels. The MOY Independent Levels will be filled using these instructional levels.

In [ ]:
# Filling Instructional Level given MOY Independent Levels
i = 1
for i in range(len(alpha_keys)-1):
    for level in alpha_keys:
        df.loc[(df['Inst Level.1']==alpha_keys[i]), 'MOY Indep Level']=alpha_keys[i-1]
        i=+1
# Checking for accuracy    
df.loc[df['MOY Indep Level'].notnull()].shape

We now have 274 complete MOY Independent Levels. I will move on to investigate other columns.

In [ ]:
df.loc[df['Grade Level Equiv.2'].isna()].shape

There are 15 records missing the MOY Grade Level Equivalency. 

In [ ]:
# Addressing Grade Level Equiv.2 for Instructional Level
for level, equiv in levels.items():
    df.loc[(df['MOY Indep Level']==level), 'Grade Level Equiv.2'] = equiv
    
# Checking for accuracy    
df.loc[df['Grade Level Equiv.2'].isna()].shape

In [ ]:
# Investigating the 7 missing records.
df.loc[df['Grade Level Equiv.2'].isna()]

The data shows that students missing MOY data also were students who were not enrolled in the beginning of the year. My assumption is that these students were registered, but didnot attend the school. I will drop these records from the data frame.

In [ ]:
df.dropna(subset=['Grade Level Equiv.2'], inplace=True)

# Checking for accuracy
df.shape

In [ ]:
df.loc[df['Inst Level.1'].isna()].shape

In [ ]:
# Investigating 10 missing records.
df.loc[df['Inst Level.1'].isna()]

In [ ]:
# Filling MOY Instructional Level given MOY Independent Levels
i = 0
for i in range(len(alpha_keys)-1):
    for level in alpha_keys:
        df.loc[(df['MOY Indep Level']==alpha_keys[i]), 'Inst Level.1']=alpha_keys[i+1]
        i=+1

# Checking for accuracy
df.loc[df['Inst Level.1'].isna()]

This first grader's BOY Independent Level was A and MOY Independent Level was Z. Such a high jump is highly unlikely. An investigation into these outliers are required.

In [ ]:
# Investigating Outliers

In [ ]:
df.info()